File: esoteric_entropy/sources/camera.py
"""Camera photon shot noise entropy source (requires opencv-python)."""

from __future__ import annotations

import numpy as np

from esoteric_entropy.sources.base import EntropySource


class CameraNoiseSource(EntropySource):
    """Entropy from camera sensor dark-current / photon shot noise.

    Image sensor pixels accumulate charge from photon arrivals (Poisson
    process) and thermal dark current.  With the lens covered, the LSBs
    of pixel values are dominated by quantum shot noise.

    Requires ``opencv-python`` optional dependency and a camera.
    """

    name = "camera_shot_noise"
    description = "Camera sensor photon shot noise / dark current"
    category = "hardware"
    physics = (
        "Captures frames from the camera sensor in darkness. The sensor's photodiodes generate dark current from thermal electron-hole pair generation in silicon — a quantum process. Read noise from the amplifier adds further randomness. The LSBs of pixel values in dark frames are dominated by shot noise (Poisson-distributed photon counting)."
    )
    platform_requirements = ["opencv-python", "camera"]
    entropy_rate_estimate = 50000.0

    def is_available(self) -> bool:
        try:
            import cv2  # noqa: F401

            cap = cv2.VideoCapture(0)
            ok = cap.isOpened()
            cap.release()
            return ok
        except Exception:
            return False

    def collect(self, n_samples: int = 10000) -> np.ndarray:
        import cv2

        cap = cv2.VideoCapture(0)
        try:
            ret, frame = cap.read()
            if not ret or frame is None:
                return np.array([], dtype=np.uint8)
            # Flatten and take LSBs
            flat = frame.flatten()
            return (flat & 0x0F).astype(np.uint8)[:n_samples]
        finally:
            cap.release()

    def entropy_quality(self) -> dict:
        data = self.collect(10000)
        return self._quick_quality(data, self.name)

File: crates/openentropy-core/src/sources/camera.rs
//! CameraNoiseSource — Camera sensor dark current noise.
//!
//! Captures a single frame from the camera via ffmpeg's avfoundation backend
//! as raw grayscale video, then extracts the lower 4 bits of each pixel value.
//! In darkness, these LSBs are dominated by shot noise and dark current.

use crate::source::{EntropySource, SourceCategory, SourceInfo};

use super::helpers::{command_exists, pack_nibbles};

static CAMERA_NOISE_INFO: SourceInfo = SourceInfo {
    name: "camera_noise",
    description: "Camera sensor dark current and shot noise via ffmpeg",
    physics: "Captures frames from the camera sensor in darkness. The sensor's photodiodes \
              generate dark current from thermal electron-hole pair generation in silicon \
              \u{2014} a quantum process. Read noise from the amplifier adds further randomness. \
              The LSBs of pixel values in dark frames are dominated by shot noise \
              (Poisson-distributed photon counting).",
    category: SourceCategory::Hardware,
    platform_requirements: &["macos"],
    entropy_rate_estimate: 50000.0,
    composite: false,
};

/// Entropy source that harvests sensor noise from camera dark frames.
pub struct CameraNoiseSource;

impl EntropySource for CameraNoiseSource {
    fn info(&self) -> &SourceInfo {
        &CAMERA_NOISE_INFO
    }

    fn is_available(&self) -> bool {
        cfg!(target_os = "macos") && command_exists("ffmpeg")
    }

    fn collect(&self, n_samples: usize) -> Vec<u8> {
        // Capture one frame of raw grayscale video from the default camera.
        // ffmpeg -f avfoundation -i "0" -frames:v 1 -f rawvideo -pix_fmt gray pipe:1
        let result = std::process::Command::new("ffmpeg")
            .args([
                "-f",
                "avfoundation",
                "-i",
                "0",
                "-frames:v",
                "1",
                "-f",
                "rawvideo",
                "-pix_fmt",
                "gray",
                "pipe:1",
            ])
            .stdin(std::process::Stdio::null())
            .stdout(std::process::Stdio::piped())
            .stderr(std::process::Stdio::null())
            .output();

        let raw_frame = match result {
            Ok(output) if output.status.success() => output.stdout,
            _ => return Vec::new(),
        };

        // Extract the lower 4 bits of each pixel value and pack nibbles.
        let nibbles = raw_frame.iter().map(|pixel| pixel & 0x0F);
        pack_nibbles(nibbles, n_samples)
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn camera_noise_info() {
        let src = CameraNoiseSource;
        assert_eq!(src.name(), "camera_noise");
        assert_eq!(src.info().category, SourceCategory::Hardware);
        assert_eq!(src.info().entropy_rate_estimate, 50000.0);
    }
}
